cpp/store_local_data.cc

```diff
@@ -0,0 +1,77 @@
+/** \brief Saves local MARC data in a database for later retrieval with the add_local_data tool.
+ *  \author Dr. Johannes Ruscheinski (johannes.ruscheinski@uni-tuebingen.de)
+ *
+ *  \copyright 2020 Universitätsbibliothek Tübingen.  All rights reserved.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <cstdio>
+#include <cstdlib>
+#include "Compiler.h"
+#include "LocalDataDB.h"
+#include "MARC.h"
+#include "StringUtil.h"
+#include "UBTools.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("marc_title_data_with_local_data");
+}
+
+
+void StoreLocalData(LocalDataDB * const local_data_db, MARC::Reader * const reader) {
+    unsigned total_record_count(0), local_data_extraction_count(0);
+    while (auto record = reader->read()) {
+        ++total_record_count;
+
+        const auto PPN(record.getControlNumber());
+
+        auto local_field(record.findTag("LOK"));
+        if (unlikely(local_field == record.end())) {
+            LOG_WARNING("records w/ PPN " + PPN + " has no local fields!");
+            continue;
+        }
+
+        std::vector<std::string> local_fields;
+        do
+            local_fields.emplace_back(local_field->getContents());
+        while (++local_field != record.end());
```

Comment:
There might be other Tags between LOK and record.end() inserted by us, depending on the phase of the pipeline (e.g. "ORI")

Reply:
True and I did consider that.  As we don't know which current and future phases need the local data I considered that we need to re-add the local data early in the pipeline and therefore it should not cause a problem.  I also added comments in the pipeline scripts.